File: 2-两数相加.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {//模拟加法，管理链表，直接执行加法似乎会超出int范围，所以解题思路是模拟加法，不要忘记最后一位进位
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        int sum = 0;
        ListNode *head = nullptr, *tail = nullptr;
        int v1 = 0, v2 = 0;
        while(l1||l2){
            v1 = l1?l1->val:0;
            v2 = l2?l2->val:0;
            sum = v1+v2+carry;
            if(!head){
                head = new ListNode(sum%10);
                tail = head;
            }
            else{
                tail->next = new ListNode(sum%10);
                tail = tail->next;
            }
            carry = sum/10;
            if(l1)
            l1 = l1->next;
            if(l2)
            l2 = l2->next;
        }
        if (carry > 0) {
            tail->next = new ListNode(carry);
        }
        return head;
    }
};

File: 2-两数相加.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode tail = null;
        int carry = 0;
        ListNode head = tail;
        while(l1 != null || l2 != null){
            int v1 = l1 == null ? 0 : l1.val;
            int v2 = l2 == null ? 0 : l2.val;
            int temp = (v1 + v2 + carry);
            ListNode t = new ListNode(temp%10,null);
            carry = temp/10;
            if(head != null){
                tail.next = t;
                tail = t;
            }
            else{
                head = t;
                tail = t;
            }
            if (l1 != null) {
                l1 = l1.next;
            }
            if (l2 != null) {
                l2 = l2.next;
            }
        }
        if(carry != 0){
            ListNode t = new ListNode(carry,null);
            tail.next = t;
            tail = t;
        }
        return head;
    }
}